RelationshipVortex - Add a high-level index to facilitate relationship queries

totten opened the pull request:
Overview
----------------------------------------

In CiviCRM, "relationships" are user-configurable relations between two contacts. The data-structure which represents these configurable relationships is complicated to query. This PR adds a supplemental data-structure (the "relationship vortex") which whirls the data around a little - and drops it off it in a more accessible spot.

Before
----------------------------------------

Each end of the relationship may be given a name (e.g. `Spouse of`, `Child of`, `Parent of`). In the `civicrm_relationship_type`, this corresponds to a few columns (*partial listing*):

* `civicrm_relationship_type`.`id`
* `civicrm_relationship_type`.`name_a_b`
* `civicrm_relationship_type`.`name_b_a`

When a specific relationship is created between two contacts, Civi creates correlated values in the `civicrm_relationship` table (*partial listing*):

* `civicrm_relationship`.`relationship_type_id`
* `civicrm_relationship`.`contact_id_a`
* `civicrm_relationship`.`contact_id_b`

The difficulty arises when building queries on this data - as a consumer of the data, I know the name of a relation that interests me (eg `Parent of`). I need to know which column (`contact_id_a` or `contact_id_b`) corresponds to each side of the relationship, and then change the query structure to match. Depending on the metadata stored in `civicrm_relationship_type`, I may need swap among `contact_id_a` and `contact_id_b` in different places - or I may need to query both ways and `UNION` the result. This grows increasingly complicated if you wish query on multiple relationships.

After
----------------------------------------

You may query `civicrm_relationship_vtx`, which has these columns (*partial listing*)::

* `near_contact_id`
* `relation`
* `far_contact_id`

As query author, you may arbitrarily decide to match your initial contact against `near_contact_id`; their counterpart will always be in `far_contact_id`; and the name in the `relation` will be set to match.

For example, if Alice (`#100`) is the parent of Bob (`#200`), then the `civicrm_relationship_vtx` would  have these two records:

| `near_contact_id`| `relation` | `far_contact_id` |
|--|--|--|
| Alice (`#100`) | `Parent of` | Bob (`#200`) |
| Bob (`#200`) | `Child of` | Alice (`#100`) |

Example
----------------------------------------

Suppose we want to lookup the immediate family for everyone subscribed to a given group (`group_id=2`). Our query should:

* Begin with group subscribers (`civicrm_group_contact`)
* Join on the "relationship"
* Filter the "relationship" type -- the immediate familial relations are "Parent of", "Child of", "Spouse of", "Sibling of".
* Join on the "contact" table to return the names of the subscribers and their relatives

Using `civicrm_relationship_vtx`, the query can be:

```php
$select = CRM_Utils_SQL_Select::from('civicrm_group_contact gc')
  ->where('gc.group_id = 2')
  ->join('relvtx', 'INNER JOIN civicrm_relationship_vtx relvtx ON (gc.contact_id = relvtx.near_contact_id AND relvtx.relation IN (@relations))', [
    '@relations' => ['Parent of', 'Child of', 'Spouse of', 'Sibling of'],
  ])
  ->join('the_subscriber', 'INNER JOIN civicrm_contact the_subscriber ON relvtx.near_contact_id = the_subscriber.id')
  ->join('the_relative', 'INNER JOIN civicrm_contact the_relative ON relvtx.far_contact_id = the_relative.id')
  ->select('the_subscriber.id as subscriber_cid, the_subscriber.display_name as subscriber_name')
  ->select('the_relative.id as relative_cid, the_relative.display_name as relative_name')
  ->select('relvtx.relation as relation')
;

echo $select->toSQL();
foreach ($select->execute()->fetchAll() as $row) {
  printf("Subscriber %s (#%d) is %s %s (#%d)\n", 
    $row['subscriber_name'], $row['subscriber_cid'], $row['relation'], $row['relative_name'], $row['relative_cid']);
}
```

Which produces output like this:

```
Subscriber Valene Blackwell (#161) is Parent of Russell Blackwell (#28)
Subscriber Russell Blackwell (#28) is Child of Valene Blackwell (#161)
Subscriber Russell Blackwell (#28) is Parent of Teddy Blackwell Sr. (#191)
Subscriber Russell Blackwell (#28) is Spouse of Brigette Bachman (#87)
```

The nice thing to note here is that we did not have do any accounting for the orientation of the relationship -- there are no unions and there is no switching between `contact_id_a` and `contact_id_b`. We start from our initial contact ID (`gc.contact_id`) and simply join the relationship's `near_contact_id`.

Technical Details
----------------------------------------

For every record in`civicrm_relationship`, there are two records in `civicrm_relationship_vtx`. (One record from the perspective of `contact_id_a`; and the other record from the perspective of `contact_id_b`.) This mapping is enforced at runtime via trigger, and it is initialized during upgrade.

Why is it called "vortex"? Well, for one thing, it swirls the data around. For another thing, the proper term for this thing would be an "index" (i.e. it is a purely derivative data-structure designed to facilitate search mechanics). But MySQL already has very specific meaning for `INDEX`, so `civicrm_relationship_index` can sound ambiguous in discussion. Vortex is a different word. (But... at least it rhymes with the right word...)

~~This PR is marked "WIP" because it needs some unit-test coverage.~~ (Added unit test)

This PR only aims to provide the data-structure. There is a follow-up need to enhance APIv4's `Relationship` support to take advantage of this mechanism.

colemanw commented:
Also need an APIv4 entity. I would suggest not extending the class `DAOEntity` and instead extend `AbstractEntity` so you can limit the actions to just `getFields` and `get`.

colemanw commented on CRM/Contact/DAO/RelationshipVortex.php:
```diff
@@ -0,0 +1,328 @@
+        'id' => [
+          'name' => 'id',
+          'type' => CRM_Utils_Type::T_INT,
+          'title' => ts('Relationship ID'),
```
I would love for this table to just not have an `id` column. It's useless and confusing. Especially with this label!

totten replied:
Agree, it's silly to have it there. My impression from @eileenmcnaughton (where she mentioned a module with a bespoke table that lacked `id`) was that it's still rather fiddly to have a table without an `id`.

The label is really confusing. Changed it.

eileenmcnaughton replied:
Note that if you don't have a primary key then replication is impacted

colemanw commented on CRM/Contact/DAO/RelationshipVortex.php:
```diff
@@ -0,0 +1,328 @@
+        'orientation' => [
+          'name' => 'orientation',
+          'type' => CRM_Utils_Type::T_INT,
+          'title' => ts('Orientation (A/B or B/A)'),
```
Existing code uses the convention `"a_b"` or `"b_a"` so it would be nice to stick with that and not have to reinvent it as `"A/B"` `"B/A"`

totten replied:
Sure, I'm pushing an update to use `a_b`/`b_a` instead of 0/1.

colemanw commented on CRM/Contact/BAO/RelationshipVortex.php:
```diff
@@ -0,0 +1,112 @@
+         * This trigger runs whenever a "civicrm_relationship_type" record is updated.
+         *
+         * Goal: Ensure that the denormalized fields ("name_b_a"/"name_a_b" <=> "type") remain current.
+         */
```
Yea I guess it's technically possible to update a relationship type name, if you force your way past the UI (which definitely doesn't allow that) and use sql. Should be very rare, but ok.

totten replied:
Yeah, it's rare, but it's more correct, and the marginal cost is very different - e.g. plugging this particular gap took maybe +15 min upfront; but any/each person who stumbles into the edge-case is liable to spend (optimistically) an hour figuring out that their UI is reporting weird data because somebody last week made an oversight during some special import/correction.

totten commented:
@colemanw - I've pushed a few updates here:

* Rebased for `5.29.alpha1` (instead of `5.28.alpha1`)
* Change the format of `orientation` to be a more familiar `a_b` or `b_a`
* Added a unit-test
* Regenerated the demo data (so that new dev sites are consistent)
* Removed "WIP" flag

For adding it into APIv4... a read-only entity sends reasonable. I can draft/post a PR for that. However, this PR (to add the data-structure) is apt to grow stale.

totten commented:
Just to verbalize one more thought that has bugged me... the phrasing of the relationship-type-names is a little counter-intuitive. Maybe it's just my English-native brain. For example, I conceive a query like this: "I want to find all of Bob's children". There is a relationship-type-name "Child of", and my brain emphasizes the word "Child", so I assume that's what I want. But that's wrong. The trailing preposition is quite significant. For the `name_a_b` or `name_b_a` to make any sense, you have to put it into a full sentence:

| `near_contact_id`| `relation` | `far_contact_id` |
|--|--|--|
| Alice (`#100`) | is `Parent of` | Bob (`#200`) |
| Bob (`#200`) | is `Child of` | Alice (`#100`) |
| Bob (`#200`) | is `Parent of` | Carol (`#100`) |
| Carol (`#300`) | is `Child of` | Bob (`#200`) |

The correct interpretation of "I want to find all of Bob's children" is:

```sql
select far_contact_id where near_contact_id=200 and relation = 'Parent of'
```

To my way of thinking, it would be most natural to have:

| `near_contact_id`| `near_role` | `far_contact_id` | `far_role` |
|--|--|--|--|
| Alice (`#100`) | `Parent` | Bob (`#200`) | `Child` |
| Bob (`#200`) | `Child` | Alice (`#100`) | `Parent` |
| Bob (`#200`) | `Parent` | Carol (`#100`) | `Child` |
| Carol (`#300`) | `Child` | Bob (`#200`) | `Parent` |

```
select far_contact_id where near_contact_id=200 and far_role = 'Child'
```

Alas, the content of `name_a_b` and `name_b_a` (and, for that matter, `label_a_b` or `label_b_a`) does not give the name of "A" and "B" roles (ie "Parent", "Child") - it gives a name of the relationship from the "A" or "B" perspective (ie "Child of", "Parent of").

In my local branches, I've gone back/forth between 3-column and 4-column. 

| `near_contact_id`| `relation` | `far_contact_id` |
|--|--|--|
| Bob (`#200`) | `Parent of` | Carol (`#100`) |

* This makes sense if you read it as a sentence ("Bob is parent of Carol")
* This is slightly more space efficient (although that doesn't really matter).
* To get what I want ("find Bob's children"), I have to mentally adapt ("Bob's children" => "people for whom Bob's relation is `Parent of`")

| `near_contact_id`| `near_relation` | `far_contact_id` | `far_relation` |
|--|--|--|--|
| Bob (`#200`) | `Parent of` | Carol (`#100`) | `Child of` |

* This is readable if you ignore the trailing prepositions.
* This is slightly less space efficient.
* To get what I want, I can filter on the `far_relation`.

Looking at this again, I rather want to flip back to the 4-column approach...

colemanw commented:
I would vote for the 3 column approach. I'm not a big fan of creating multiple ways of doing the same thing unless we're doing it for backward-compat reasons, and this is greenfield.

demeritcowboy commented:
I see the note about delete cascade, and I'm not sure I want to get involved here, but how does this work with is_active and start/end date? Those maybe don't really apply for Parent/Child too much, and maybe this PR is just about relationships like that?

Would you want triggers that remove from the vortex when is_active becomes 0 or end-date passes, and then re-adds when those change?

Similarly in the upgrade do you want to exclude those from population?

I don't know if I need to give an example but

name_a_b: is Starship Captain of
name_b_a: has as Starship Captain

* Pike is Starship Captain of The Enterprise, start date (stardate) 1217.6
* Then some stuff happened and that relationship ended, let's say on date 1847.3
* Kirk becomes Starship Captain of The Enterprise, starting let's say 2036.2

So when you query who the Enterprise's captain is, if you don't take start/end or is_active into account you get two results.

Then this could get even more complicated because eventually Kirk was promoted to Admiral and so his relationship with the Enterprise ended. Then, after he saved the earth from a planet-killing spaceprobe created by humpback whales (I'm not making that up), he was demoted because he also did some bad stuff along the way and so he is now Captain again, with a new start date. You'd get (at least) 3 search results.

colemanw commented:
@demeritcowboy raises some excellent points, although his example does beg the question of data modeling; would a contact type "Starship" be a sub-type of "Organization" or "Household"? It's kind of both if you think about it...

That aside, I think the question of is_active is an excellent one to raise here, since the purpose of this table is to make relationships easy to query. By normalizing the relationship direction, we're halfway there, and the second-most annoying part of querying the relationship table is that to retrieve current relationships you need to consider 3 columns: `is_active`, `start_date` and `end_date`. A typical query for current relationships looks like `WHERE is_active AND (start_date IS NULL OR start_date <= NOW()) AND (end_date IS NULL OR end_date > NOW())`

I propose we add a column to this table called `is_current` and it would be a synthesis of the above. In order to keep it up-to-date we would need a trigger on each of the 3 aforementioned columns. We would also need to employ a "temporal trigger" (aka [MySql scheduled event](https://www.mysqltutorial.org/mysql-triggers/working-mysql-scheduled-event/)) to run at midnight each day to update the `is_current` field for active relationships that have a start or end date.

totten commented:
God bless unit tests - the test failure showed an faulty interaction between the logging-trigger and vortex-trigger. 391ffee  resolved it locally.

totten commented:
@demeritcowboy @colemanw I agree it'd be easier to query if it mirrors more information, and I'm happy to turn it from a thinner mirror to a thicker mirror. Of course, as coleman noted before, there is also the critique about having multiple ways to do the same thing - and mirroring more means that there are more ways to do the same thing.

IMHO, we can choose between these designs:

1. _Minimal_: The vortex table should have the minimum number of fields required to get over the a/b orientation problem.
    * That's currently what the branch does -- the "minimum" includes elements needed to identify rows and propagate updates (`id`, `relationship_id`, `relationship_type_id`, `orientation`) - and then columns which decode the a/b orientation (ie `near_contact_id`, `relation`, `far_contact_id`.)
    * In a minimal design, you'd have to `JOIN` to lookup the `start_date`/`end_date`/etc
2. _Maximal_: Mirror most of the `civicrm_relationship` to `civicrm_relationship_vtx` so that it's unnecessary to join them. (Thus, we'd also mirror `start_date`, `end_date`, `_is_active`, `description`, and probably some variation on `is_permission_a_b`/`is_permission_b_a` and possibly `case_id`.)

The `is_current` flag is interesting - because I strongly agree that it would be easier to query if it was a well-defined column, and the only issues are with implementation. I hadn't heard of MySQL "events" before, but that sounds really cool and (on its own intrinsic merits) probably the best way to compute `is_current`. I think we could make it fairly efficient (w/suitable indexes and only performing writes when there's a change), and the sysadmin costs would be fine for one instance. But I am concerned about the systemic costs for downstream sysadmins:

* We've heard several times about how folks on shared/managed/iaas-type systems have difficulty updating `mysqld`, and this is not a common `mysqld` option. (See, eg, [these hoops](https://stackoverflow.com/a/19714599/4195300) required on Amazon's MySQL.) 
* I'm unclear how it works with replication. ([This](https://dev.mysql.com/doc/refman/5.7/en/replication-features-invoked.html) makes it sound like you have to be rather painstaking to coordinate the activation/deactivation of events on different nodes.)

Other thoughts on `is_current`:

* The `is_current` stuff does not depend on A/B orientation - it could just as well be tracked in `civicrm_relationship`. (*The main reason why `civicrm_relationship_vtx` needs to be a separate table is that grain/multiplicity is different.*)
* While I don't entirely love the reliability of `cron`... using `cron` to toggle `civicrm_relationship.is_current` would be consistent with how `civicrm_membership` and `civicrm_mailing` records are scheduled.
* If the value of `is_current` is stored, then it can be indexed (hooray!)
* If the value of `is_current` is stored, then it must be calculated with a singular timezone. This is a change -- IMHO, an improvement -- over how `is_current` is interpreted with an on-demand query. (When doing on-demand queries with `DATE`s, the computations are usually subjective - dependent upon the viewer's timezone. I'd wager that Meg in Baltimore would occasionally see the same `Relationship` differently from Tom in Seattle. If the `Relationship` is set to begin on Valentine's Day, then Meg will realize the start about 3 hours before Tom realizes it. To avoid confusion, you need either `TIMESTAMP`s or a canonical timezone.)

Overall, I'd sort of prefer to have `is_current` as a separate issue. However, if we're gonna head in that direction, then I'm happy to refactor a bit so that it's easier to mirror the additional columns.

colemanw commented:
I agree with punting the `is_current` issue as something we can do a little later.
Another approach for it would be to add sugar to the api that will generate the complex query with a simple param.

colemanw commented on xml/schema/Contact/RelationshipVortex.xml:
```diff
@@ -0,0 +1,132 @@
+    <type>int unsigned</type>
+    <title>Contact ID (Far side)</title>
+    <required>true</required>
+    <comment>id of the second contact</comment>
+    <add>5.29</add>
+    <html>
+      <type>EntityRef</type>
+    </html>
```
```suggestion
    <comment>id of the related contact</comment>
    <add>5.29</add>
```

colemanw commented:
@totten can you squash some of these commits?

totten commented:
@colemanw @demeritcowboy  - Yup, I rebased and squashed the previous 15 commits into ~7 commits (331a3d85ab956f86).

But... I've also done some refactoring to make it easier to maintain the mappings (which leads to some more commits/rebasing/sqaushing - now  598c1c2). The previous mix of queries were a bit stringy -- i.e. very similar, but slightly different, and reproduced multiple times. That's fine if the table is minimalist (ie not changing much), but it got a bit tedious to juggle as we went through PR revisions.

So the updated version extracts a helper method (`$select->syncInto('tgt_table'...)` aka #17780) which makes it much easier to read+revise. Note the use of `CRM_Contact_BAO_RelationshipVortex::$mappings` which determines the triggers as well as the `rebuild()` logic.

eileenmcnaughton commented:
I've had a quick read & I'll think through it more however 2 quick things

1) it's stale
2) I can accept the idea that it is a relationship vortex - but then abbreviating vortex to vtx really just descends into meaninglessness IMHO

colemanw commented:
Closing in favor of #17781
